File: src/commands/dev/tests/metrics.rs
//! 测试指标相关命令
//!
//! 提供测试指标收集等功能。

use crate::base::util::file::{FileReader, FileWriter};
use color_eyre::{eyre::WrapErr, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::{log_info, log_success, log_break};

/// 测试报告摘要
#[derive(Debug, Clone, Default, Deserialize, Serialize)]
struct ReportSummary {
    #[serde(default)]
    total: usize,
    #[serde(default)]
    passed: usize,
    #[serde(default)]
    failed: usize,
    #[serde(default)]
    ignored: usize,
    #[serde(default)]
    timeout: usize,
    #[serde(default)]
    success_rate: f64,
    #[serde(default)]
    duration_secs: f64,
}

/// 测试用例
#[derive(Debug, Clone, Deserialize, Serialize)]
struct TestCase {
    name: String,
    #[serde(default)]
    module: String,
    #[serde(default)]
    status: String,
    #[serde(default)]
    duration_secs: f64,
}

/// 测试报告
#[derive(Debug, Clone, Deserialize, Serialize)]
struct TestReport {
    #[serde(default)]
    summary: ReportSummary,
    #[serde(default)]
    test_cases: Vec<TestCase>,
}

/// 模块统计
#[derive(Debug, Clone, Serialize)]
struct ModuleStats {
    total: usize,
    passed: usize,
    failed: usize,
    ignored: usize,
    timeout: usize,
    total_duration: f64,
    failure_rate: f64,
    success_rate: f64,
}

/// 指标数据
#[derive(Debug, Serialize)]
struct MetricsData {
    timestamp: String,
    summary: ReportSummary,
    module_stats: HashMap<String, ModuleStats>,
    test_cases: Vec<TestCase>,
    #[serde(skip_serializing_if = "Option::is_none")]
    metadata: Option<Metadata>,
}

/// 元数据
#[derive(Debug, Clone, Serialize)]
struct Metadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    test_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    platform: Option<String>,
}

/// 测试指标收集命令
pub struct TestsMetricsCollectCommand {
    report: Option<String>,
    output: Option<String>,
}

impl TestsMetricsCollectCommand {
    /// 创建新的测试指标收集命令
    pub fn new(report: Option<String>, output: Option<String>) -> Self {
        Self { report, output }
    }

    /// 收集测试指标
    pub fn collect(&self) -> Result<()> {
        let report_path = self.report.as_ref()
            .ok_or_else(|| color_eyre::eyre::eyre!("--report 参数是必需的"))?;
        let output_path = self.output.as_ref()
            .ok_or_else(|| color_eyre::eyre::eyre!("--output 参数是必需的"))?;

        log_break!('=');
        log_info!("收集测试指标");
        log_break!('=');
        log_break!();

        // 读取测试报告
        log_info!("读取测试报告: {}", report_path);
        let reader = FileReader::new(report_path);
        let report: TestReport = reader.json()
            .wrap_err_with(|| format!("Failed to parse test report: {}", report_path))?;

        // 收集指标
        log_info!("收集指标数据...");
        let metrics = self.collect_metrics_from_report(&report)?;

        // 保存指标文件
        log_info!("保存指标文件: {}", output_path);
        let writer = FileWriter::new(output_path);
        let json_content = serde_json::to_string_pretty(&metrics)
            .wrap_err("Failed to serialize metrics")?;
        writer.write_str_with_dir(&json_content)
            .wrap_err_with(|| format!("Failed to write metrics file: {}", output_path))?;

        log_break!();
        log_success!("指标收集完成");
        log_info!("   总测试数: {}", metrics.summary.total);
        log_info!("   通过: {}", metrics.summary.passed);
        log_info!("   失败: {}", metrics.summary.failed);
        log_info!("   成功率: {:.2}%", metrics.summary.success_rate);
        log_info!("   模块数: {}", metrics.module_stats.len());
        log_break!();

        Ok(())
    }

    /// 从测试报告收集指标
    fn collect_metrics_from_report(&self, report: &TestReport) -> Result<MetricsData> {
        let summary = &report.summary;
        let test_cases = &report.test_cases;

        // 按模块统计
        let mut module_stats: HashMap<String, ModuleStats> = HashMap::new();

        for test in test_cases {
            let module = test.module.clone();
            let stats = module_stats.entry(module).or_insert_with(|| ModuleStats {
                total: 0,
                passed: 0,
                failed: 0,
                ignored: 0,
                timeout: 0,
                total_duration: 0.0,
                failure_rate: 0.0,
                success_rate: 0.0,
            });

            stats.total += 1;
            stats.total_duration += test.duration_secs;

            match test.status.as_str() {
                "Passed" => stats.passed += 1,
                "Failed" => stats.failed += 1,
                "Ignored" => stats.ignored += 1,
                "Timeout" => stats.timeout += 1,
                _ => {}
            }
        }

        // 计算失败率和成功率
        for stats in module_stats.values_mut() {
            if stats.total > 0 {
                stats.failure_rate = (stats.failed as f64 / stats.total as f64) * 100.0;
                stats.success_rate = (stats.passed as f64 / stats.total as f64) * 100.0;
            }
        }

        // 生成时间戳
        let timestamp = chrono::Utc::now().to_rfc3339();

        // 构建指标数据
        let metrics = MetricsData {
            timestamp,
            summary: summary.clone(),
            module_stats,
            test_cases: test_cases.clone(),
            metadata: None,
        };

        Ok(metrics)
    }
}

File: scripts/dev/py/tests/metrics.py
"""测试指标收集实现"""
import json
import sys
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from typing import Dict, List

# 添加父目录到路径
sys.path.insert(0, str(Path(__file__).parent.parent))
from utils.logger import log_info, log_success, log_break


def collect(args) -> None:
    """收集测试指标"""
    if not args.report:
        print("Error: --report 参数是必需的", file=sys.stderr)
        sys.exit(1)
    if not args.output:
        print("Error: --output 参数是必需的", file=sys.stderr)
        sys.exit(1)

    log_break('=')
    log_info("收集测试指标")
    log_break('=')
    log_break()

    # 读取测试报告
    log_info(f"读取测试报告: {args.report}")
    try:
        with open(args.report, 'r') as f:
            report = json.load(f)
    except Exception as e:
        print(f"Error: Failed to parse test report: {e}", file=sys.stderr)
        sys.exit(1)

    # 收集指标
    log_info("收集指标数据...")
    metrics = _collect_metrics_from_report(report)

    # 保存指标文件
    log_info(f"保存指标文件: {args.output}")
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    with open(output_path, 'w') as f:
        json.dump(metrics, f, indent=2, ensure_ascii=False)

    log_break()
    log_success("指标收集完成")
    summary = metrics.get('summary', {})
    log_info(f"   总测试数: {summary.get('total', 0)}")
    log_info(f"   通过: {summary.get('passed', 0)}")
    log_info(f"   失败: {summary.get('failed', 0)}")
    log_info(f"   成功率: {summary.get('success_rate', 0.0):.2f}%")
    log_info(f"   模块数: {len(metrics.get('module_stats', {}))}")
    log_break()


def _collect_metrics_from_report(report: Dict) -> Dict:
    """从测试报告收集指标"""
    summary = report.get('summary', {})
    test_cases = report.get('test_cases', [])

    # 按模块统计
    module_stats: Dict[str, Dict] = defaultdict(lambda: {
        'total': 0,
        'passed': 0,
        'failed': 0,
        'ignored': 0,
        'timeout': 0,
        'total_duration': 0.0,
        'failure_rate': 0.0,
        'success_rate': 0.0
    })

    for test in test_cases:
        module = test.get('module', 'unknown')
        stats = module_stats[module]

        stats['total'] += 1
        stats['total_duration'] += test.get('duration_secs', 0.0)

        status = test.get('status', '')
        if status == 'Passed':
            stats['passed'] += 1
        elif status == 'Failed':
            stats['failed'] += 1
        elif status == 'Ignored':
            stats['ignored'] += 1
        elif status == 'Timeout':
            stats['timeout'] += 1

    # 计算失败率和成功率
    for stats in module_stats.values():
        if stats['total'] > 0:
            stats['failure_rate'] = (stats['failed'] / stats['total']) * 100.0
            stats['success_rate'] = (stats['passed'] / stats['total']) * 100.0

    # 生成时间戳
    timestamp = datetime.utcnow().isoformat() + 'Z'

    # 构建指标数据
    metrics = {
        'timestamp': timestamp,
        'summary': summary.copy(),
        'module_stats': dict(module_stats),
        'test_cases': test_cases.copy(),
        'metadata': None
    }

    return metrics


def main():
    """CLI 入口（可以直接运行此脚本）"""
    import argparse

    parser = argparse.ArgumentParser(
        prog='metrics',
        description='Collect test metrics'
    )
    parser.add_argument('--report', required=True, help='Test report file path')
    parser.add_argument('--output', required=True, help='Output metrics file path')

    args = parser.parse_args()
    collect(args)


if __name__ == '__main__':
    main()

